diff --git a/node_modules/@dfinity/bls-verify/lib/cjs/index.js b/node_modules/@dfinity/bls-verify/lib/cjs/index.js
index 35fb7b7..afa8e6e 100644
--- a/node_modules/@dfinity/bls-verify/lib/cjs/index.js
+++ b/node_modules/@dfinity/bls-verify/lib/cjs/index.js
@@ -1,7 +1,7 @@
 "use strict";
 Object.defineProperty(exports, "__esModule", { value: true });
 exports.blsVerify = void 0;
-const amcl_js_1 = require("amcl-js");
+const amcl_js_1 = require("./vendor/amcl-js");
 const utils_1 = require("./utils");
 const blsVerify = async (primaryKey, signature, message) => {
     const pk = typeof primaryKey === 'string' ? primaryKey : (0, utils_1.toHex)(primaryKey);
